BrewR-EPFL/BrewR
# feat: Implement offline mode and real-time updates for gallery screen

**This pull request implements offline mode functionality and real-time updates for the gallery screen, enhancing the user experience both online and offline. Below are the details of the changes:**

Key Features Implemented:
Offline Coffee Shop Selection:

Users can now input a coffee shop name as plain text if no dropdown suggestions are available.
This is particularly useful for offline mode, where location suggestions might not load.
Offline Journey Addition with Default Image:

When adding a journey offline, a placeholder image is displayed.
Once the device reconnects to the internet, the actual image is uploaded, and the gallery is automatically updated.
Real-Time Updates in Gallery:

Updated the getJourneys method in JourneysRepositoryFirestore to use Firestore's addSnapshotListener instead of get().
Ensures that gallery changes are reflected in real time, whether online or offline.
Network Connectivity Check for Image Upload:

Introduced a new utility class, CheckConnection.kt, to verify internet connectivity before attempting image uploads.
Prevents unnecessary upload attempts when offline and ensures smooth synchronization when back online.
Modified Files:
**AddJourney.kt:**

Added logic to handle offline mode with a default placeholder image.
Incorporated connectivity checks before image uploads.
Ensured synchronization of journeys when the device reconnects to the internet.
**JourneyComponents.kt:**

Updated UI components to gracefully handle and display placeholder images in offline mode.
**JourneysRepositoryFirestore.kt:**

Replaced get() with addSnapshotListener in the getJourneys method to enable real-time updates for the gallery.
**CheckConnection.kt (New):**

Added a utility class to check network connectivity before performing operations.
How to Test:
Offline Mode:

Add a journey offline and verify that the placeholder image is displayed in the gallery.
Reconnect to the internet and confirm that the gallery updates with the uploaded image.
Real-Time Updates:

Add or edit a journey online and verify that the gallery updates in real time without refreshing.
Coffee Shop Selection:

Add a journey and input a coffee shop name manually when dropdown suggestions are unavailable.
Connectivity Check:

Disable the internet and attempt to upload an image.
Confirm that the app uses the default placeholder image instead of trying to upload.

**Screenshot**

![offlinemode photo](https://github.com/user-attachments/assets/fd499604-acf3-47c2-b098-70437111180c)

Users can add journey in offline mode


![ChangeImplementedOnline](https://github.com/user-attachments/assets/09706496-c95a-4212-a06c-f91985cd5780)
The photo that the user uploaded is shown when the internet connection is back
